src/stash/src/postgres.rs

```diff
@@ -0,0 +1,470 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Durable metadata storage.
+
+use std::cmp;
+use std::marker::PhantomData;
+
+use postgres::{Client, IsolationLevel, Transaction};
+use timely::progress::Antichain;
+use timely::PartialOrder;
+
+use mz_persist_types::Codec;
+use timely::progress::frontier::AntichainRef;
+
+use crate::{
+    AntichainFormatter, Diff, Id, InternalStashError, Stash, StashCollection, StashError, Timestamp,
+};
+
+const SCHEMA: &str = "
+CREATE TABLE fence (
+    version bigint PRIMARY KEY
+);
+INSERT INTO fence VALUES (1);
+
+CREATE TABLE collections (
+    collection_id bigserial PRIMARY KEY,
+    name text NOT NULL UNIQUE
+);
+
+CREATE TABLE data (
+    collection_id bigint NOT NULL REFERENCES collections (collection_id),
+    key bytea NOT NULL,
+    value bytea NOT NULL,
+    time bigint NOT NULL,
+    diff bigint NOT NULL
+);
+
+CREATE INDEX data_time_idx ON data (collection_id, time);
+
+CREATE TABLE sinces (
+    collection_id bigint NOT NULL UNIQUE REFERENCES collections (collection_id),
+    since bigint
+);
+
+CREATE TABLE uppers (
+    collection_id bigint NOT NULL UNIQUE REFERENCES collections (collection_id),
+    upper bigint
+);
+";
+
+/// A Stash whose data is stored in a single file on disk. The format of this
+/// file is not specified and should not be relied upon. The only promise is
+/// stability. Any changes to the on-disk format will be accompanied by a clear
+/// migration path.
+pub struct Postgres {
+    conn: Client,
+    fence: i64,
+}
+
+// Start a transaction for later use with commit_transaction. These are macros
+// instead of functions due to rust borrowing complaints. TODO: Is there a way
+// to implement these as regular functions?
+macro_rules! start_transaction {
+    ($self:expr, $tx:ident) => {
+        let mut $tx = $self
+            .conn
+            .build_transaction()
+            // Ensure serialized fence checking.
+            .isolation_level(IsolationLevel::Serializable)
+            .start()?;
+    };
+}
```

Comment:
Could instead set `default_transaction_isolation` on the connection when you create it.

Reply:
Alternatively, just make it a free function that takes a `Conn`.

Reply:
Actually, even better: make a `transact` function which takes a closure:

```diff
diff --git a/src/stash/src/postgres.rs b/src/stash/src/postgres.rs
index 5443ea9b4..19e9bb115 100644
--- a/src/stash/src/postgres.rs
+++ b/src/stash/src/postgres.rs
@@ -64,43 +64,39 @@ pub struct Postgres {
     fence: i64,
 }
 
-// Start a transaction for later use with commit_transaction. These are macros
-// instead of functions due to rust borrowing complaints. TODO: Is there a way
-// to implement these as regular functions?
-macro_rules! start_transaction {
-    ($self:expr, $tx:ident) => {
-        let mut $tx = $self
+impl Postgres {
+    /// Construct a fenced transaction.
+    fn transact<F, T>(&mut self, f: F) -> Result<T, StashError>
+    where
+        F: FnOnce(&mut Transaction) -> Result<T, StashError>,
+    {
+        let mut tx = self
             .conn
             .build_transaction()
             // Ensure serialized fence checking.
             .isolation_level(IsolationLevel::Serializable)
             .start()?;
-    };
-}
 
-// Commit a transaction, verifying and updating the fence version from the
-// server. This needs to be a macro instead of a normal function call because
-// of drop ordering with the conn.
-macro_rules! commit_transaction {
-    ($self:expr, $tx:ident) => {
-        let current: i64 = $tx.query_one("SELECT version FROM fence", &[])?.get(0);
-        if current != $self.fence {
+        let res = f(&mut tx)?;
+
+        let current: i64 = tx.query_one("SELECT version FROM fence", &[])?.get(0);
+        if current != self.fence {
             return Err(InternalStashError::Fence(format!(
                 "unexpected fence {}, expected {}",
-                current, $self.fence
+                current, self.fence
             ))
             .into());
         }
-        let next = $self.fence + 1;
-        $tx.execute("UPDATE fence SET version = $1", &[&next])?;
-        $tx.commit()?;
+        let next = self.fence + 1;
+        tx.execute("UPDATE fence SET version = $1", &[&next])?;
+        tx.commit()?;
         // Only after we successfully commit the updated fence is the in-memory fence
         // updated.
-        $self.fence = next;
-    };
-}
+        self.fence = next;
+
+        Ok(res)
+    }
 
-impl Postgres {
     /// Opens the stash stored at the specified path.
     pub fn open(mut conn: Client) -> Result<Postgres, StashError> {
         let mut tx = conn.transaction()?;
@@ -195,41 +191,40 @@ impl Stash for Postgres {
         K: Codec + Ord,
         V: Codec + Ord,
     {
-        start_transaction!(self, tx);
+        self.transact(|tx| {
+            let collection_id_opt: Option<_> = tx
+                .query_one(
+                    "SELECT collection_id FROM collections WHERE name = $1",
+                    &[&name],
+                )
+                .map(|row| row.get("collection_id"))
+                .ok();
 
-        let collection_id_opt: Option<_> = tx
-            .query_one(
-                "SELECT collection_id FROM collections WHERE name = $1",
-                &[&name],
-            )
-            .map(|row| row.get("collection_id"))
-            .ok();
+            let collection_id = match collection_id_opt {
+                Some(id) => id,
+                None => {
+                    let collection_id = tx
+                        .query_one(
+                            "INSERT INTO collections (name) VALUES ($1) RETURNING collection_id",
+                            &[&name],
+                        )?
+                        .get("collection_id");
+                    tx.execute(
+                        "INSERT INTO sinces (collection_id, since) VALUES ($1, $2)",
+                        &[&collection_id, &Timestamp::MIN],
+                    )?;
+                    tx.execute(
+                        "INSERT INTO uppers (collection_id, upper) VALUES ($1, $2)",
+                        &[&collection_id, &Timestamp::MIN],
+                    )?;
+                    collection_id
+                }
+            };
 
-        let collection_id = match collection_id_opt {
-            Some(id) => id,
-            None => {
-                let collection_id = tx
-                    .query_one(
-                        "INSERT INTO collections (name) VALUES ($1) RETURNING collection_id",
-                        &[&name],
-                    )?
-                    .get("collection_id");
-                tx.execute(
-                    "INSERT INTO sinces (collection_id, since) VALUES ($1, $2)",
-                    &[&collection_id, &Timestamp::MIN],
-                )?;
-                tx.execute(
-                    "INSERT INTO uppers (collection_id, upper) VALUES ($1, $2)",
-                    &[&collection_id, &Timestamp::MIN],
-                )?;
-                collection_id
-            }
-        };
-
-        commit_transaction!(self, tx);
-        Ok(StashCollection {
-            id: collection_id,
-            _kv: PhantomData,
+            Ok(StashCollection {
+                id: collection_id,
+                _kv: PhantomData,
+            })
         })
     }
 
@@ -241,34 +236,35 @@ impl Stash for Postgres {
         K: Codec + Ord,
         V: Codec + Ord,
     {
-        start_transaction!(self, tx);
-        let since = match Self::since_tx(&mut tx, collection.id)?.into_option() {
-            Some(since) => since,
-            None => {
-                return Err(StashError::from(
-                    "cannot iterate collection with empty since frontier",
-                ));
-            }
-        };
-        let mut rows = tx
-            .query(
-                "SELECT key, value, time, diff FROM data
-                 WHERE collection_id = $1",
-                &[&collection.id],
-            )?
-            .into_iter()
-            .map(|row| {
-                let key_buf: Vec<_> = row.try_get("key")?;
-                let value_buf: Vec<_> = row.try_get("value")?;
-                let key = K::decode(&key_buf)?;
-                let value = V::decode(&value_buf)?;
-                let time = row.try_get("time")?;
-                let diff = row.try_get("diff")?;
-                Ok::<_, StashError>(((key, value), cmp::max(time, since), diff))
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-        differential_dataflow::consolidation::consolidate_updates(&mut rows);
-        Ok(rows)
+        self.transact(|tx| {
+            let since = match Self::since_tx(tx, collection.id)?.into_option() {
+                Some(since) => since,
+                None => {
+                    return Err(StashError::from(
+                        "cannot iterate collection with empty since frontier",
+                    ));
+                }
+            };
+            let mut rows = tx
+                .query(
+                    "SELECT key, value, time, diff FROM data
+                    WHERE collection_id = $1",
+                    &[&collection.id],
+                )?
+                .into_iter()
+                .map(|row| {
+                    let key_buf: Vec<_> = row.try_get("key")?;
+                    let value_buf: Vec<_> = row.try_get("value")?;
+                    let key = K::decode(&key_buf)?;
+                    let value = V::decode(&value_buf)?;
+                    let time = row.try_get("time")?;
+                    let diff = row.try_get("diff")?;
+                    Ok::<_, StashError>(((key, value), cmp::max(time, since), diff))
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            differential_dataflow::consolidation::consolidate_updates(&mut rows);
+            Ok(rows)
+        })
     }
 
     fn iter_key<K, V>(
@@ -282,32 +278,33 @@ impl Stash for Postgres {
     {
         let mut key_buf = vec![];
         key.encode(&mut key_buf);
-        start_transaction!(self, tx);
-        let since = match Self::since_tx(&mut tx, collection.id)?.into_option() {
-            Some(since) => since,
-            None => {
-                return Err(StashError::from(
-                    "cannot iterate collection with empty since frontier",
-                ));
-            }
-        };
-        let mut rows = tx
-            .query(
-                "SELECT value, time, diff FROM data
-                 WHERE collection_id = $1 AND key = $2",
-                &[&collection.id, &key_buf],
-            )?
-            .into_iter()
-            .map(|row| {
-                let value_buf: Vec<_> = row.try_get("value")?;
-                let value = V::decode(&value_buf)?;
-                let time = row.try_get("time")?;
-                let diff = row.try_get("diff")?;
-                Ok::<_, StashError>((value, cmp::max(time, since), diff))
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-        differential_dataflow::consolidation::consolidate_updates(&mut rows);
-        Ok(rows)
+        self.transact(|tx| {
+            let since = match Self::since_tx(tx, collection.id)?.into_option() {
+                Some(since) => since,
+                None => {
+                    return Err(StashError::from(
+                        "cannot iterate collection with empty since frontier",
+                    ));
+                }
+            };
+            let mut rows = tx
+                .query(
+                    "SELECT value, time, diff FROM data
+                    WHERE collection_id = $1 AND key = $2",
+                    &[&collection.id, &key_buf],
+                )?
+                .into_iter()
+                .map(|row| {
+                    let value_buf: Vec<_> = row.try_get("value")?;
+                    let value = V::decode(&value_buf)?;
+                    let time = row.try_get("time")?;
+                    let diff = row.try_get("diff")?;
+                    Ok::<_, StashError>((value, cmp::max(time, since), diff))
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            differential_dataflow::consolidation::consolidate_updates(&mut rows);
+            Ok(rows)
+        })
     }
 
     fn update_many<K: Codec, V: Codec, I>(
@@ -325,10 +322,7 @@ impl Stash for Postgres {
             value.encode(&mut value_buf);
             ((key_buf, value_buf), time, diff)
         });
-        start_transaction!(self, tx);
-        Self::update_many_tx(&mut tx, collection.id, entries)?;
-        commit_transaction!(self, tx);
-        Ok(())
+        self.transact(|tx| Self::update_many_tx(tx, collection.id, entries))
     }
 
     fn seal<K, V>(
@@ -346,10 +340,7 @@ impl Stash for Postgres {
         let seals = seals
             .iter()
             .map(|(collection, frontier)| (collection.id, frontier));
-        start_transaction!(self, tx);
-        Self::seal_batch_tx(&mut tx, seals)?;
-        commit_transaction!(self, tx);
-        Ok(())
+        self.transact(|tx| Self::seal_batch_tx(tx, seals))
     }
 
     fn compact<K, V>(
@@ -364,29 +355,29 @@ impl Stash for Postgres {
         &mut self,
         compactions: &[(StashCollection<K, V>, Antichain<Timestamp>)],
     ) -> Result<(), StashError> {
-        start_transaction!(self, tx);
-        let compact_stmt = tx.prepare("UPDATE sinces SET since = $1 WHERE collection_id = $2")?;
-        for (collection, new_since) in compactions {
-            let since = Self::since_tx(&mut tx, collection.id)?;
-            let upper = Self::upper_tx(&mut tx, collection.id)?;
-            if PartialOrder::less_than(&upper, new_since) {
-                return Err(StashError::from(format!(
-                    "compact request {} is greater than the current upper frontier {}",
-                    AntichainFormatter(new_since),
-                    AntichainFormatter(&upper)
-                )));
+        self.transact(|tx| {
+            let compact_stmt = tx.prepare("UPDATE sinces SET since = $1 WHERE collection_id = $2")?;
+            for (collection, new_since) in compactions {
+                let since = Self::since_tx(tx, collection.id)?;
+                let upper = Self::upper_tx(tx, collection.id)?;
+                if PartialOrder::less_than(&upper, new_since) {
+                    return Err(StashError::from(format!(
+                        "compact request {} is greater than the current upper frontier {}",
+                        AntichainFormatter(new_since),
+                        AntichainFormatter(&upper)
+                    )));
+                }
+                if PartialOrder::less_than(new_since, &since) {
+                    return Err(StashError::from(format!(
+                        "compact request {} is less than the current since frontier {}",
+                        AntichainFormatter(new_since),
+                        AntichainFormatter(&since)
+                    )));
+                }
+                tx.execute(&compact_stmt, &[&new_since.as_option(), &collection.id])?;
             }
-            if PartialOrder::less_than(new_since, &since) {
-                return Err(StashError::from(format!(
-                    "compact request {} is less than the current since frontier {}",
-                    AntichainFormatter(new_since),
-                    AntichainFormatter(&since)
-                )));
-            }
-            tx.execute(&compact_stmt, &[&new_since.as_option(), &collection.id])?;
-        }
-        commit_transaction!(self, tx);
-        Ok(())
+            Ok(())
+        })
     }
 
     fn consolidate<K, V>(&mut self, collection: StashCollection<K, V>) -> Result<(), StashError> {
@@ -397,45 +388,45 @@ impl Stash for Postgres {
         &mut self,
         collections: &[StashCollection<K, V>],
     ) -> Result<(), StashError> {
-        start_transaction!(self, tx);
+        self.transact(|tx| {
+            let consolidation_stmt = tx.prepare(
+                "DELETE FROM data
+                WHERE collection_id = $1 AND time <= $2
+                RETURNING key, value, diff",
+            )?;
+            let insert_stmt = tx.prepare(
+                "INSERT INTO data (collection_id, key, value, time, diff)
+                VALUES ($1, $2, $3, $4, $5)",
+            )?;
+            let drop_stmt = tx.prepare("DELETE FROM data WHERE collection_id = $1")?;
 
-        let consolidation_stmt = tx.prepare(
-            "DELETE FROM data
-             WHERE collection_id = $1 AND time <= $2
-             RETURNING key, value, diff",
-        )?;
-        let insert_stmt = tx.prepare(
-            "INSERT INTO data (collection_id, key, value, time, diff)
-             VALUES ($1, $2, $3, $4, $5)",
-        )?;
-        let drop_stmt = tx.prepare("DELETE FROM data WHERE collection_id = $1")?;
-
-        for collection in collections {
-            let since = Self::since_tx(&mut tx, collection.id)?.into_option();
-            match since {
-                Some(since) => {
-                    let mut updates = tx
-                        .query(&consolidation_stmt, &[&collection.id, &since])?
-                        .into_iter()
-                        .map(|row| {
-                            let key = row.try_get("key")?;
-                            let value = row.try_get("value")?;
-                            let diff = row.try_get("diff")?;
-                            Ok::<_, StashError>(((key, value), since, diff))
-                        })
-                        .collect::<Result<Vec<((Vec<u8>, Vec<u8>), i64, i64)>, _>>()?;
-                    differential_dataflow::consolidation::consolidate_updates(&mut updates);
-                    for ((key, value), time, diff) in updates {
-                        tx.execute(&insert_stmt, &[&collection.id, &key, &value, &time, &diff])?;
+            for collection in collections {
+                let since = Self::since_tx(tx, collection.id)?.into_option();
+                match since {
+                    Some(since) => {
+                        let mut updates = tx
+                            .query(&consolidation_stmt, &[&collection.id, &since])?
+                            .into_iter()
+                            .map(|row| {
+                                let key = row.try_get("key")?;
+                                let value = row.try_get("value")?;
+                                let diff = row.try_get("diff")?;
+                                Ok::<_, StashError>(((key, value), since, diff))
+                            })
+                            .collect::<Result<Vec<((Vec<u8>, Vec<u8>), i64, i64)>, _>>()?;
+                        differential_dataflow::consolidation::consolidate_updates(&mut updates);
+                        for ((key, value), time, diff) in updates {
+                            tx.execute(&insert_stmt, &[&collection.id, &key, &value, &time, &diff])?;
+                        }
+                    }
+                    None => {
+                        tx.execute(&drop_stmt, &[&collection.id])?;
                     }
                 }
-                None => {
-                    tx.execute(&drop_stmt, &[&collection.id])?;
-                }
             }
-        }
-        commit_transaction!(self, tx);
-        Ok(())
+
+            Ok(())
+        })
     }
 
     /// Reports the current since frontier.
@@ -443,10 +434,7 @@ impl Stash for Postgres {
         &mut self,
         collection: StashCollection<K, V>,
     ) -> Result<Antichain<Timestamp>, StashError> {
-        start_transaction!(self, tx);
-        let since = Self::since_tx(&mut tx, collection.id)?;
-        commit_transaction!(self, tx);
-        Ok(since)
+        self.transact(|tx| Self::since_tx(tx, collection.id))
     }
 
     /// Reports the current upper frontier.
@@ -454,10 +442,7 @@ impl Stash for Postgres {
         &mut self,
         collection: StashCollection<K, V>,
     ) -> Result<Antichain<Timestamp>, StashError> {
-        start_transaction!(self, tx);
-        let upper = Self::upper_tx(&mut tx, collection.id)?;
-        commit_transaction!(self, tx);
-        Ok(upper)
+        self.transact(|tx| Self::upper_tx(tx, collection.id))
     }
 }
 
```